libs/utils/src/http/endpoint.rs

```diff
@@ -63,9 +87,53 @@ async fn prometheus_metrics_handler(_req: Request<Body>) -> Result<Response<Body
     Ok(response)
 }
 
+pub fn add_request_id_middleware<B: hyper::body::HttpBody + Send + Sync + 'static>(
+) -> Middleware<B, ApiError> {
+    Middleware::pre(move |req| async move {
+        let request_id = match req.headers().get(&X_REQUEST_ID_HEADER) {
+            Some(request_id) => {
+                Cow::Borrowed(request_id.to_str().expect("extract request id value"))
+            }
+            None => {
+                let request_id = uuid::Uuid::new_v4();
+                Cow::Owned(request_id.to_string())
+            }
+        };
+
+        req.set_context(RequestId {
+            id: request_id.to_string(),
+        });
+
+        if req.method() == Method::GET {
+            tracing::debug!("{} {} {}", req.method(), req.uri().path(), request_id);
+        } else {
+            tracing::info!("{} {} {}", req.method(), req.uri().path(), request_id);
+        }
+        Ok(req)
+    })
+}
+
+async fn add_request_id_header_to_response(
+    mut res: Response<Body>,
+    req_info: RequestInfo,
+) -> Result<Response<Body>, ApiError> {
+    if let Some(request_id) = req_info.context::<RequestId>() {
```

Comment:
Thinking about defensive checks here. Not sure we should add them, maybe debug_assert or something. Should be fine as is too

Reply:
we don't want to fail if its missing